src/libcore/slice/mod.rs

```diff
@@ -2405,8 +2421,21 @@ macro_rules! iterator {
 
             #[inline]
             fn nth(&mut self, n: usize) -> Option<$elem> {
-                // Call helper method. Can't put the definition here because mut versus const.
-                self.iter_nth(n)
+                if n >= self.len() {
+                    // This iterator is now empty.  The way we encode the length of a non-ZST
+                    // iterator, this works for both ZST and non-ZST.
+                    // For a ZST we would usually do `self.end = self.ptr`, but since
+                    // we will not give out an address any more after this there is no
+                    // way to observe the difference.
```

Comment:
Couldn't I still get an address from `.as_slice().as_ptr()`?

Reply:
Hm... okay it is observable but only as a raw pointer, not as a reference. So there's no soundness problem.

I can still change it, if you prefer.

Reply:
Up to you; the comment tweak is probably fine since I'm not sure why anyone would be looking at the address of a post-iteration slice anyway.